File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func minDepth(root *TreeNode) int {

	var dfs func(*TreeNode, int) int
	dfs = func(treeNode *TreeNode, index int) int {
		if treeNode.Left == nil && treeNode.Right == nil {
			return index
		}

		var left, right int
		var leftValid, rightValid bool
		if treeNode.Left != nil {
			leftValid = true
			left = dfs(treeNode.Left, index+1)
		}
		if treeNode.Right != nil {
			rightValid = true
			right = dfs(treeNode.Right, index+1)
		}

		if !leftValid {
			return right
		}
		if !rightValid {
			return left
		}

		if left < right {
			return left
		} else {
			return right
		}
	}

    if root == nil {
        return 0
    }
    
	return dfs(root, 1)

}

File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return dfs(root);
    }

    private int dfs(TreeNode node) {
        if (node == null) {
            return Integer.MAX_VALUE;
        }
        
        if (node.left == null && node.right == null) {
            return 1;
        }

        int leftDepth = dfs(node.left);
        int rightDepth = dfs(node.right);

        return Math.min(leftDepth, rightDepth) + 1;
    }
}